optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/forager/finalist/AbstractFinalistPodium.java

```diff
@@ -44,6 +46,19 @@ public void stepStarted(LocalSearchStepScope stepScope) {
         finalistList = new ArrayList<>(1024);
     }
 
+    protected void clearAndAddFinalist(LocalSearchMoveScope moveScope) {
+        finalistList.clear();
+        finalistList.add(moveScope);
+    }
+
+    protected void addFinalist(LocalSearchMoveScope moveScope) {
+        if (finalistList.size() >= FINALIST_LIST_MAX_SIZE) {
+            // Avoid unbounded growth and OutOfMemoryException
+            return;
```

Comment:
Wouldn't it make sense to instead remove the least recent finalist? Or perhaps the worst finalist?

Reply:
All moves in the finalistList have the same score (which is the reason that in most cases, it's typically size 1 to at most 100). There is no way to distinguish out the worst one.

I doubted about ignoring the oldest or the newest.
Removing the oldest in an ArrayList (not ArrayDeque) has a big0(n) perf cost. Ignoring the newest doesn't.
Both the first move evaluated and the last move evaluated in the same step with the same score, have equal right to be selected - recency doesn't count (there is no changes in the steps).
So I went for ignoring the newest.